centralise rose cylc docs (#1033)
ACCESS-Hive/access-hive.github.io: docs/models/run_a_model/rose_cylc.md

    <terminal-line>Please enter the MOSRS password for &lt;MOSRS-username&gt;:</terminal-line>
    <terminal-line lineDelay=1500>Checking your credentials using Subversion. Please wait.</terminal-line>
    <terminal-line lineDelay=500><span style="color: #559cd5;">INFO</span>: Successfully accessed Subversion with your credentials.</terminal-line>
    <terminal-line lineDelay=100><span style="color: #559cd5;">INFO</span>: Checking your credentials using rosie. Please wait.</terminal-line>
    <terminal-line lineDelay=500><span style="color: #559cd5;">INFO</span>: Successfully accessed rosie with your credentials.</terminal-line>
</terminal-window>

## Get the model configuration

Depending on the the specific model, its configuration will be hosted either on _Github_ or _MOSRS_. The [Run a Model](https://docs.access-hive.org.au/models/run-a-model/) documentation for the specific model will specify where the configuration is stored.<br>

Regardless of where the configuration comes from, it is recommended to store the local copy in the `~/roses/` (this happens automatically for configurations pulled from MOSRS).

> ### Model configurations stored on Github

Author: I've added instructions on what to do if you want to make play around/make exploratory changes to Github configurations, and have them tracked.

The newly created persistent session is assigned a unique identifier, referred to here as `<persistent-session-uuid>`.

### Assign the persistent session to Cylc {: .no-toc }

Once the session is running, it needs to be assigned to _Cylc_. This is done by inserting the persistent session label into `~/.persistent-sessions/cylc-session`, which can be done with the following command (substituting `<name>` and `<project>` for the name and project used to create the persistent session).

```
cat > ~/.persistent-sessions/cylc-session <<< "<name>.${USER}.<project>.ps.gadi.nci.org.au"
```

You can check that this worked with:

```
cat ~/.persistent-sessions/cylc-session
```

For example, if user `abc123` started a persistent session named `CM2` under the project `tm70`, then the command would be:

<terminal-window data="input">
    <terminal-line>cat > ~/.persistent-sessions/cylc-session <<< CM2.abc123.tm70.ps.gadi.nci.org.au</terminal-line>
    <terminal-line data="input" linedelay="1000">cat ~/.persistent-sessions/cylc-session</terminal-line>
    <terminal-line data="output">CM2.abc123.tm70.ps.gadi.nci.org.au</terminal-line>

> </terminal-window>

Author: I must've marked this as resolved by accident. Has now been addressed.

# Run models using Rose/Cylc

## About

> The _Rose/Cylc_ workflow management tool consists of two components: the [_Cylc_](https://niwa.co.nz/environmental-information/cylc-suite-engine) task engine, developed by the New Zealand National Institute of Water and Atmospheric Research (NIWA), and the [_Rose_](https://www.metoffice.gov.uk/research/approach/modelling-systems/rose) framework developed by the UK Met Office (UKMO) which configures tasks for the _Cylc_ engine. 

Reviewer: The _Rose/Cylc_ workflow management tool consists of two components: 
  - a [_Cylc_](https://niwa.co.nz/environmental-information/cylc-suite-engine) task engine developed by the New Zealand National Institute of Water and Atmospheric Research (NIWA), and 
  - a [_Rose_](https://www.metoffice.gov.uk/research/approach/modelling-systems/rose) framework developed by the UK Met Office (UKMO) which configures tasks for the _Cylc_ engine.

A set of tasks configured by _Rose_ to run with the _Cylc_ engine is called a _suite_ (for _Cylc 7_) or _workflow_ (for _Cylc 8_).

!!! warning

>     ACCESS models configurations that run with _Rose/Cylc_ use _Cylc7_, with intentions to upgrade to _Cylc8_. The upgrade is expected to change some aspects of the workflow described on this page. Updated information about the _Cylc8_ workflow will be provided once a model configuration using this version becomes available.

Reviewer: _Edit:_
"ACCESS models configurations **are** run with _Rose/Cylc_ use _Cylc7_,..."

Reviewer 2: I don't think your suggestion is correct, did you mean:
"ACCESS models configurations are run with Rose/Cylc using Cylc7 ..."?

In any case I would stick with the current text.

## Prerequisites

- **NCI account**<br> 
    Before running an ACCESS model, you need to [Set Up your NCI Account](/getting_started/set_up_nci_account).

- **_MOSRS_ account**<br>

>     The [Met Office Science Repository Service (MOSRS)](https://code.metoffice.gov.uk) is a server run by the UK Met Office (UKMO) to support collaborative development with other partners organisations. MOSRS contains the source code for some ACCESS model components and configurations, and a MOSRS account is a license requirement to run any ACCESS those configurations.<br>

Reviewer: _Edit:_ you have already defined UKMO so not needed again here
**"... the UKMO to support..."**

Reviewer 2: Thank for picking this up Heidi. I agree we should be consistent with the style we write MOSRS in. 
Since MOSRS is an acronym, I would use simple style here. Italic I would say might be better used for proper nouns (specific tools or names, e.g., _Cylc_, _Gadi_, etc.)

    To apply for a _MOSRS_ account, please contact your [local institutional sponsor](https://opus.nci.org.au/display/DAE/Prerequisites).
    {: #mosrs-account}

- **Join NCI projects**<br>
    Join the following projects by requesting membership on their respective NCI project pages:

    - [access](https://my.nci.org.au/mancini/project/access/join)
    - [hr22](https://my.nci.org.au/mancini/project/hr22/join)
    
## Connecting to Gadi

You can run _Rose/Cylc_ either from a _Gadi_ login node, or via an [ARE VDI session](https://opus.nci.org.au/spaces/Help/pages/163250532/2.1.+Connecting+to+the+VDI). If you wish to use the Gadi login node, skip directly to [Set up a persistent session](#set-up-a-persistent-session).

<div markdown id="x11-forwarding">
!!! warning "X11 Forwarding"

>     If you are going to connect via the terminal directly using SSH, it is recommended to connect with `ssh -X` to enable X11 forwarding. This allows the Rose and Cylc GUIs to be launched on your local machine.

Reviewer: _Edit:_
"**If using a terminal to connect directly using SSH**, it is recommended to connect with `ssh -X` to enable X11 forwarding. This allows the **_Rose_ and _Cylc_** GUIs to be launched on your local machine."

<terminal-window data="input">
    <terminal-line>cat > ~/.persistent-sessions/cylc-session <<< CM2.abc123.tm70.ps.gadi.nci.org.au</terminal-line>
    <terminal-line data="input" linedelay="1000">cat ~/.persistent-sessions/cylc-session</terminal-line>
    <terminal-line data="output">CM2.abc123.tm70.ps.gadi.nci.org.au</terminal-line>
</terminal-window>

### List active persistent sessions {: .no-toc }

To list currently active sessions, use:

```
persistent-sessions list
```
### Terminate a persistent sessions {: .no-toc }

To end a specific session, use:

```
persistent-sessions kill <persistent-session-uuid>
```

## Rose/Cylc Setup

### Rose and Cylc executables

Make the `rose` and `cylc` executables available by loading the _Cylc_ module:

```
module use /g/data/hr22/modulefiles
module load cylc<version>
```

where `<version>` is the version of _Cylc_ used by the respective configuration.

### MOSRS Authentication

The ACCESS models which use _Cylc_ require a connection to the MOSRS mirror on Gadi. To connect to this mirror, you must first authenticate your MOSRS credentials with:

```
mosrs-auth
```

> This will request the username and password you received when you created your MOSRS account.

Reviewer 2: Same as above

<terminal-window>
    <terminal-line data="input">mosrs-auth</terminal-line>
    <terminal-line lineDelay=500><span style="color: #559cd5;">INFO</span>: You need to enter your MOSRS credentials here so that GPG can cache your password.</terminal-line>
    <terminal-line>Please enter the MOSRS password for &lt;MOSRS-username&gt;:</terminal-line>
    <terminal-line lineDelay=1500>Checking your credentials using Subversion. Please wait.</terminal-line>
    <terminal-line lineDelay=500><span style="color: #559cd5;">INFO</span>: Successfully accessed Subversion with your credentials.</terminal-line>
    <terminal-line lineDelay=100><span style="color: #559cd5;">INFO</span>: Checking your credentials using rosie. Please wait.</terminal-line>
    <terminal-line lineDelay=500><span style="color: #559cd5;">INFO</span>: Successfully accessed rosie with your credentials.</terminal-line>
</terminal-window>

## Get the model configuration

> Depending on the the specific model, its configuration will be hosted either on _Github_ or _MOSRS_. The [Run a Model](/models/run-a-model/) documentation for the respective model will specify where the configuration is stored.<br>

Reviewer 2: This is the correct link:
```suggestion
Depending on the the specific model, its configuration will be hosted either on _Github_ or _MOSRS_. The [Run a Model](/models/run_a_model/) documentation for the respective model will specify where the configuration is stored.<br>
```

> Regardless of where the configuration comes from, it is recommended to store the local copy in the `~/roses/` directory (this happens automatically for configurations pulled from MOSRS).

Reviewer: _Edit:_
"... pulled from _MOSRS_)."

Reviewer 2: Same as above

### Model configurations stored on Github

For Github hosted configurations, get a local copy by cloning the Github repository with:

```
git -C ~/roses clone <repository> -b <branch>
```

> where `<repository>` and `<branch>` are specific to the chosen model configuration and can be found in the respective [Run a Model documentation](/models/run-a-model/).

Reviewer: _Edit:_ To be consistent with identical reference above:
"... [Run a Model](/models/run-a-model/) documentation." not "[Run a Model documentation](/models/run-a-model/)"
_Broken link:_ (/models/run-a-model/)

If you want to make exploratory changes within the configuration, and have those changes tracked, please [fork the configuration repository](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/fork-a-repo) and commit your changes there.<br>

> If you think your new configuration provide significant value to the broader community, refer to the respective model configuration documentation for instructions on how to have it officially supported by ACCESS-NRI.

Reviewer: _Typo:_
"... provides..."

### Model configurations stored on MOSRS

> There are two ways of getting a local copy of a configuration hosted on MOSRS:

Reviewer: _Edit:_
"... hosted on _MOSRS_:"

- [Local-only copy](#rosie-checkout)
- [Local and remote copy (new remote configuration)](#rosie-copy)

#### Local-only copy {: #rosie-checkout }

```
rosie checkout <suite-id>/<branch>
```

where `<suite-id>` and `<branch>` are specific to the chosen model configuration and can be found in the respective [Run a Model documentation](/models/run-a-model/). This creates a local copy of the configuration, which is placed in the `~/roses/<suite-id>` folder.

!!! tip
    To copy from the default branch (`trunk`), omit the `/<branch>` portion of the command.

Configurations obtained in this way cannot be pushed back to the remote, so use of this command is recommended for testing and examining configurations.

#### Local and remote copy (new remote configuration) {: #rosie-copy }

Before creating a new remote copy of the configuration, please read [these guidelines](https://code.metoffice.gov.uk/trac/roses-u/) on what should be stored in the Rosie repository.

```
rosie copy <suite-id>/<branch>
```

> where `<suite-id>` and `<branch>` are specific to the chosen model configuration and can be found in the respective [Run a Model documentation](/models/run-a-model/).

Reviewer: Edit: To be consistent with identical references above:
"... [Run a Model](https://github.com/models/run-a-model/) documentation." not "[Run a Model documentation](https://github.com/models/run-a-model/)"

Reviewer 2: ```suggestion
where `<suite-id>` and `<branch>` are specific to the chosen model configuration and can be found in the respective [Run a Model](/models/run_a_model/) documentation.
```

```

## Run the model configuration

To run the configuration, execute the following command from within the configuration directory:

```
rose suite-run
```

This launches the _Rose/Cylc_ configuration and opens the _Cylc_ GUI (if you are running on the login node and the GUI doesn't appear, make sure you connected with [X11 forwarding](#x11-forwarding) enabled). If you closed the GUI but want to re-open it, navigate to the configuration directory and run:

```
rose suite-gcontrol &
```

!!! tip
    The `&` is optional. It detaches the invoked process, allowing the terminal prompt to remain active while the GUI is open.

By default, the configuration, log files and outputs are copied to `/scratch/<project>/${USER}/cylc-run/<suite-id>`. See the respective model documentation for details on what outputs are generated and where to find them.

## Edit the model configuration

To edit the model configuration, run the following command from the configuration directory:

```
rose edit &
```

This opens the _Rose_ GUI that contains information about the configuration settings. The configurable options for each model configuration are described in their respective documentation pages.

> Once settings have been modified in the _Rose_ GUI, save them by clicking on the _Save_ button ![Save button](/assets/run_access_cm/save_button.png){: style="height:1em"}.

Reviewer 2: How are you imagining the sentence? Wouldn't the sentence "The configurable options for each model configuration are described in their respective documentation pages." already direct users to the model-specific information?

> Once settings have been modified in the _Rose_ GUI, save them by clicking on the _Save_ button ![Save button](/assets/run_access_cm/save_button.png){: style="height:1em"}.

Reviewer 3: I think what Heidi is pointing to is that the sentence doesn't tell the users where that documentation is, only that apparently something exists.
 
So should we point the users to the ACCESS Models page on ACCESS-Hive? Just to give users an indication of where to go to find the docs? Could add:

... their respective documentation pages (see each Model documentation page (link here to ACCESS Models) for access to their configuration documentation when it exists).

> Once settings have been modified in the _Rose_ GUI, save them by clicking on the _Save_ button ![Save button](/assets/run_access_cm/save_button.png){: style="height:1em"}.

Reviewer 2: I see now. Yes, it would be good to add a reference to a model documentation page. As Claire suggested, I think the best option we have for now is a link to the Hive Docs' Models page.

> Once settings have been modified in the _Rose_ GUI, save them by clicking on the _Save_ button ![Save button](/assets/run_access_cm/save_button.png){: style="height:1em"}.

Reviewer 2: As Claire mentioned, what is missing is a link to a page. I would link the `/models/access_models` page.

> Once settings have been modified in the _Rose_ GUI, save them by clicking on the _Save_ button ![Save button](/assets/run_access_cm/save_button.png){: style="height:1em"}.

Reviewer: A link would be good. When I wrote that comment, I was thinking "where would we like the user to be directed once they finished running through this rose/cylc setup?" Ending the section with saving the rose/cylc settings seemed unfinished to me.

> Once settings have been modified in the _Rose_ GUI, save them by clicking on the _Save_ button ![Save button](/assets/run_access_cm/save_button.png){: style="height:1em"}.

Reviewer 2: I agree, at the end a link to Run a model makes more sense to me.

## Prerequisites

- **NCI account**<br> 
    Before running an ACCESS model, you need to [Set Up your NCI Account](/getting_started/set_up_nci_account).

- **MOSRS account**<br>
    The [Met Office Science Repository Service (MOSRS)](https://code.metoffice.gov.uk) is a server run by the UKMO to support collaborative development with other partners organisations. MOSRS contains the source code for some ACCESS model components and configurations, and a MOSRS account is a license requirement to run any ACCESS those configurations.<br>
    To apply for a MOSRS account, please contact your [local institutional sponsor](https://opus.nci.org.au/display/DAE/Prerequisites).
    {: #mosrs-account}

- **Join NCI projects**<br>
    Join the following projects by requesting membership on their respective NCI project pages:

    - [access](https://my.nci.org.au/mancini/project/access/join)
    - [hr22](https://my.nci.org.au/mancini/project/hr22/join)
    
## Connecting to Gadi

> You can run _Rose/Cylc_ either from a _Gadi_ login node, or via an [ARE VDI session](https://opus.nci.org.au/spaces/Help/pages/163250532/2.1.+Connecting+to+the+VDI). If you wish to use the Gadi login node, skip directly to [Set up a persistent session](#set-up-a-persistent-session).

Reviewer 3: ```suggestion
You can run _Rose/Cylc_ either from a _Gadi_ login node, or via an [ARE VDI session](https://opus.nci.org.au/spaces/Help/pages/163250532/2.1.+Connecting+to+the+VDI). If you wish to use the _Gadi_ login node, skip directly to [Set up a persistent session](#set-up-a-persistent-session).
```

It is recommended to only have one active persistent session at any one time as several _Cylc_ sessions can use the same persistent session.

Note that persistent sessions are terminated during the quarterly maintenance at NCI and will need to be restarted afterwards. The new persistent session can be given the same name as used previously.

### Start a new persistent session

Start a new persistent session by running:

```
persistent-sessions start -p <project> <name>
```

where `<project>` is the project you want to start the session under, and `<name>` is the name you want to give your persistent session. 

!!! warning
    Persistent session names accept only a limited set of characters. We recommend using only alpha-numeric characters without spaces or underscores.

<terminal-window data="input">
    <terminal-line>persistent-sessions -p &lt;project&gt; start &lt;name&gt;</terminal-line>
    <terminal-line data="output">session &lt;persistent-session-uuid&gt; running - connect using</terminal-line>
    <terminal-line data="output">&emsp;ssh &lt;name&gt;.&lt;$USER&gt;.&lt;project&gt;.ps.gadi.nci.org.au</terminal-line>
</terminal-window>

!!! tip

>     If `-p <project>` is ommitted, your default project `$PROJECT` will be used.

Reviewer 3: ```suggestion
    If `-p <project>` is ommitted, your default project `$PROJECT` will be used.
    
    Persistent sessions can run simulations using compute and storage resources from any project, independently of the project used for the persistent session itself.
```

I feel this isn't completely clear and people might think to "play it safe" by running several persistent sessions under several projects.

```
cat > ~/.persistent-sessions/cylc-session <<< "<name>.${USER}.<project>.ps.gadi.nci.org.au"
```

You can check that this worked with:

```
cat ~/.persistent-sessions/cylc-session
```

For example, if user `abc123` started a persistent session named `CM2` under the project `tm70`, then the command would be:

<terminal-window data="input">
    <terminal-line>cat > ~/.persistent-sessions/cylc-session <<< CM2.abc123.tm70.ps.gadi.nci.org.au</terminal-line>
    <terminal-line data="input" linedelay="1000">cat ~/.persistent-sessions/cylc-session</terminal-line>
    <terminal-line data="output">CM2.abc123.tm70.ps.gadi.nci.org.au</terminal-line>
</terminal-window>

### List active persistent sessions {: .no-toc }

To list currently active sessions, use:

```
persistent-sessions list
```
### Terminate a persistent sessions {: .no-toc }

To end a specific session, use:

```
persistent-sessions kill <persistent-session-uuid>
```

## Rose/Cylc Setup

### Rose and Cylc executables

Make the `rose` and `cylc` executables available by loading the _Cylc_ module:

```
module use /g/data/hr22/modulefiles
module load cylc<version>
```

> where `<version>` is the version of _Cylc_ used by the respective configuration.

Reviewer 2: Maybe it will. Let's worry about it when the time comes.

### MOSRS Authentication

> The ACCESS models which use _Cylc_ require a connection to the MOSRS mirror on Gadi. To connect to this mirror, you must first authenticate your MOSRS credentials with:

Reviewer 3: ```suggestion
The ACCESS models which use _Cylc_ require a connection to the MOSRS mirror on _Gadi_. To connect to this mirror, you must first authenticate your MOSRS credentials with:
```

```
mosrs-auth
```

This will request the username and password you received when you created your MOSRS account.

<terminal-window>
    <terminal-line data="input">mosrs-auth</terminal-line>
    <terminal-line lineDelay=500><span style="color: #559cd5;">INFO</span>: You need to enter your MOSRS credentials here so that GPG can cache your password.</terminal-line>
    <terminal-line>Please enter the MOSRS password for &lt;MOSRS-username&gt;:</terminal-line>
    <terminal-line lineDelay=1500>Checking your credentials using Subversion. Please wait.</terminal-line>
    <terminal-line lineDelay=500><span style="color: #559cd5;">INFO</span>: Successfully accessed Subversion with your credentials.</terminal-line>
    <terminal-line lineDelay=100><span style="color: #559cd5;">INFO</span>: Checking your credentials using rosie. Please wait.</terminal-line>
    <terminal-line lineDelay=500><span style="color: #559cd5;">INFO</span>: Successfully accessed rosie with your credentials.</terminal-line>
</terminal-window>

> 

Reviewer 3: We don't say the authentication step has to be done everyday. It only talks here about the first authentication. Or will this information be on the Run a Model page?

We could add:

```suggestion
This will request the username and password you received when you created your MOSRS account.

<terminal-window>
    <terminal-line data="input">mosrs-auth</terminal-line>
    <terminal-line lineDelay=500><span style="color: #559cd5;">INFO</span>: You need to enter your MOSRS credentials here so that GPG can cache your password.</terminal-line>
    <terminal-line>Please enter the MOSRS password for &lt;MOSRS-username&gt;:</terminal-line>
    <terminal-line lineDelay=1500>Checking your credentials using Subversion. Please wait.</terminal-line>
    <terminal-line lineDelay=500><span style="color: #559cd5;">INFO</span>: Successfully accessed Subversion with your credentials.</terminal-line>
    <terminal-line lineDelay=100><span style="color: #559cd5;">INFO</span>: Checking your credentials using rosie. Please wait.</terminal-line>
    <terminal-line lineDelay=500><span style="color: #559cd5;">INFO</span>: Successfully accessed rosie with your credentials.</terminal-line>
</terminal-window>

After the first authentication, you will need to run `mosrs-auth` every 24 hours and for every new connection to _Gadi_ (e.g. new terminal) to verify your password against the saved credentials.
```

Author: Yea, probably worth including this.

## Get the model configuration

> Depending on the the specific model, its configuration will be hosted either on _Github_ or MOSRS. The [Run a Model](/models/run_a_model/) documentation for the respective model will specify where the configuration is stored.<br>

Author: I think this should actually be italicised for consistency with the other names we use e.g. *Gadi*, *Rose*, *Cylc*?

Regardless of where the configuration comes from, it is recommended to store the local copy in the `~/roses/` directory (this happens automatically for configurations pulled from MOSRS).

### Model configurations stored on Github

For Github hosted configurations, get a local copy by cloning the Github repository with:

> 

Reviewer 3: ```suggestion
### Model configurations stored on GitHub

For GitHub hosted configurations, get a local copy by cloning the GitHub repository with:

```

    <terminal-line>[INFO] &lt;new-suite-id&gt;: local copy created at &lt;$HOME&gt;/roses/&lt;new-suite-id&gt;</terminal-line>
</terminal-window>

This creates a new remote configuration with a new `suite_id` (based off the copied configuration) clones a copy of it locally in the `~/roses/<new-suite-id>` folder. Configurations created in this way are separate from the original copied configuration and can be modified and pushed back to the remote.

To push a configuration back to the remote, from within the configuration directory run:

```
fcm commit
```

## Run the model configuration

To run the configuration, execute the following command from within the configuration directory:

```
rose suite-run
```

This launches the _Rose/Cylc_ configuration and opens the _Cylc_ GUI (if you are running on the login node and the GUI doesn't appear, make sure you connected with [X11 forwarding](#x11-forwarding) enabled). If you closed the GUI but want to re-open it, navigate to the configuration directory and run:

```
rose suite-gcontrol &
```

!!! tip
    The `&` is optional. It detaches the invoked process, allowing the terminal prompt to remain active while the GUI is open.

> By default, the configuration, log files and outputs are copied to `/scratch/<project>/${USER}/cylc-run/<suite-id>`. See the respective model documentation for details on what outputs are generated and where to find them.

Reviewer 2: Again, this would be mostly the actual config docs, but for now we don't have a page to link for that, so I guess the access models page is good enough.

## Edit the model configuration

To edit the model configuration, run the following command from the configuration directory:

```
rose edit &
```

> This opens the _Rose_ GUI that contains information about the configuration settings. The configurable options for each model configuration are described in their respective documentation pages.

Reviewer 3: ```suggestion
This opens the _Rose_ GUI that contains information about the configuration settings. The configurable options for each model configuration are described in their respective [documentation pages](https://docs.access-hive.org.au/models/access_models/).
```

Reviewer 3: The OM3 config docs are linked from the OM3 model page. That's what I used to decide. I wasn't sure whether we would want a meta config docs page.

Reviewer 3: I think any high level option that appear in the Run a Model page should also be in the configs docs. Because the config docs should be a complete docs of the available options.

Reviewer 2: Yes I agree. In general most config docs are still babies as they've been around for just a few weeks, but with more time and development I hope they will be the main point of reference for a specific model.

Reviewer 2: When I talk about ACCESS models page, I refer to the [ACCESS Models page](https://docs.access-hive.org.au/models/access_models/) 😀

Author: I've made a minor change to the wording of this section after discussion with @heidinett. The current `Run a model` pages have descriptions of some simple configuration options, that are the things we'd like a beginner to play around with first. This means we should link to the `Run a Model` page here, and allow the respective `Run a model` pages then link to that model's configuration documentation which includes a more detailed description of the options.

The pages linked from the [ACCESS Models page](https://docs.access-hive.org.au/models/access_models/) don't say anything about the configuration options, and shouldn't be linked in this specific section.

# Run models using Rose/Cylc

## About
The _Rose/Cylc_ workflow management tool consists of two components:

* The [_Cylc_](https://niwa.co.nz/environmental-information/cylc-suite-engine) task engine, developed by the New Zealand National Institute of Water and Atmospheric Research (NIWA)
* The [_Rose_](https://www.metoffice.gov.uk/research/approach/modelling-systems/rose) framework developed by the UK Met Office (UKMO) which configures tasks for the _Cylc_ engine. 

> A set of tasks configured by _Rose_ to run with the _Cylc_ engine is called a _suite_ (for _Cylc 7_) or _workflow_ (for _Cylc 8_).

Reviewer 2: ```suggestion
A set of tasks configured by _Rose_ to run with the _Cylc 7_ engine is called a _suite_.
```

- **MOSRS account**<br>
    The [Met Office Science Repository Service (MOSRS)](https://code.metoffice.gov.uk) is a server run by the UKMO to support collaborative development with other partners organisations. MOSRS contains the source code for some ACCESS model components and configurations, and a MOSRS account is a license requirement to run some ACCESS configurations.<br>
    To apply for a MOSRS account, please contact your [local institutional sponsor](https://opus.nci.org.au/display/DAE/Prerequisites).
    {: #mosrs-account}

- **Join NCI projects**<br>
    Join the following projects by requesting membership on their respective NCI project pages:

    - [hr22](https://my.nci.org.au/mancini/project/hr22/join)
    
## Connecting to Gadi

You can run _Rose/Cylc_ either from a _Gadi_ login node, or via an [ARE VDI session](https://opus.nci.org.au/spaces/Help/pages/163250532/2.1.+Connecting+to+the+VDI). If you wish to use the _Gadi_ login node, skip directly to [Set up a persistent session](#set-up-a-persistent-session).

<div markdown id="x11-forwarding">
!!! warning "X11 Forwarding"

>     If using a terminal to connect directly using SSH, it is recommended to connect with `ssh -X` to enable X11 forwarding. This allows the _Rose_ and _Cylc_ GUIs to be launched on your local machine.

Reviewer 2: ```suggestion
    When connecting via SSH from a terminal, it is recommended to enable [X11 forwarding](https://some-natalie.dev/blog/ssh-x11-forwarding/), for example by adding the -X option to the `ssh` command.
```

- **Walltime (hours)** &rarr; `2`<br>
    Amount of hours the ARE VDI session will remain active for. This is only setup time, and does not reflect how long the actual configuration will take to run.

- **Queue** &rarr; `normalbw`
    
- **Compute Size** &rarr; `tiny` (1 CPU)<br>

- **Project** &rarr; a project of which you are a member.<br>
    The project must have allocated [_Service Units (SU)_](https://opus.nci.org.au/spaces/Help/pages/236881132/Allocations...) to run your simulation. By default, this will be set to your default project `$PROJECT`.

- **Storage** &rarr; `gdata/hr22, scratch/$PROJECT` (minimum)<br>
    The storage folders listed above are the minimum required to run _Rose/Cylc_.

Once the ARE VDI session opens in your browser, click the terminal icon at the top of the window to open a terminal. Use this terminal for all subsequent steps in this guide.

## Set up a persistent session

NCI provides a service called [_persistent sessions_](https://opus.nci.org.au/spaces/Help/pages/241926895/Persistent+Sessions) to enable long running processes, like _Cylc_, to stay active even when the user disconnects from _Gadi_.

> It is recommended to only have one active persistent session at any one time as several _Cylc_ sessions can use the same persistent session.

Reviewer 2: ```suggestion
It is recommended to have only one active persistent session at any given time, as multiple _Cylc_ sessions can use the same persistent session.
```

```
persistent-sessions start -p <project> <name>
```

where `<project>` is the project you want to start the session under, and `<name>` is the name you want to give your persistent session. 

!!! warning
    Persistent session names accept only a limited set of characters. We recommend using only alpha-numeric characters without spaces or underscores.

<terminal-window data="input">
    <terminal-line>persistent-sessions -p &lt;project&gt; start &lt;name&gt;</terminal-line>
    <terminal-line data="output">session &lt;persistent-session-uuid&gt; running - connect using</terminal-line>
    <terminal-line data="output">&emsp;ssh &lt;name&gt;.&lt;$USER&gt;.&lt;project&gt;.ps.gadi.nci.org.au</terminal-line>
</terminal-window>

!!! tip
    If `-p <project>` is ommitted, your default project `$PROJECT` will be used.

Persistent sessions can run simulations using compute and storage resources from any project, independently of the project used for the persistent session itself. The newly created persistent session is assigned a unique identifier, referred to here as `<persistent-session-uuid>`.

### Assign the persistent session to Cylc {: .no-toc }

> Once the session is running, it needs to be assigned to _Cylc_. This is done by inserting the persistent session label into `~/.persistent-sessions/cylc-session`, which can be done with the following command (substituting `<name>` and `<project>` for the name and project used to create the persistent session).

Reviewer 2: ```suggestion
Once the session is running, it needs to be assigned to _Cylc_. This is done by inserting the persistent session label into `~/.persistent-sessions/cylc-session`, which can be done with the following command (substituting `<name>` and `<project>` with the name and project used to create the persistent session).
```

The ACCESS models which use _Cylc_ require a connection to the MOSRS mirror on _Gadi_. To connect to this mirror, you must first authenticate your MOSRS credentials with:

```
mosrs-auth
```

This will request the username and password you received when you created your MOSRS account.

<terminal-window>
    <terminal-line data="input">mosrs-auth</terminal-line>
    <terminal-line lineDelay=500><span style="color: #559cd5;">INFO</span>: You need to enter your MOSRS credentials here so that GPG can cache your password.</terminal-line>
    <terminal-line>Please enter the MOSRS password for &lt;MOSRS-username&gt;:</terminal-line>
    <terminal-line lineDelay=1500>Checking your credentials using Subversion. Please wait.</terminal-line>
    <terminal-line lineDelay=500><span style="color: #559cd5;">INFO</span>: Successfully accessed Subversion with your credentials.</terminal-line>
    <terminal-line lineDelay=100><span style="color: #559cd5;">INFO</span>: Checking your credentials using rosie. Please wait.</terminal-line>
    <terminal-line lineDelay=500><span style="color: #559cd5;">INFO</span>: Successfully accessed rosie with your credentials.</terminal-line>
</terminal-window>

> After the first authentication, you will need to run `mosrs-auth` every 24 hours and for every new connection to _Gadi_ (e.g. new terminal) to verify your password against the saved credentials.

Reviewer 2: ```suggestion
After the first authentication, you will need to run `mosrs-auth` every 24 hours and for every new connection to _Gadi_ (e.g., new terminal) to verify your password against the saved credentials.
```

Author: Are you sure this is right? That looks grammatically incorrect to me

Reviewer 2: We're always using `e.g.,`.
https://ontariotraining.net/grammar-tip-punctuation-with-i-e-and-e-g/

If you want to make exploratory changes within the configuration, and have those changes tracked, please [fork the configuration repository](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/fork-a-repo) and commit your changes there.<br>
If you think your new configuration provides significant value to the broader community, refer to the respective model configuration documentation for instructions on how to have it officially supported by ACCESS-NRI.

### Model configurations stored on MOSRS

There are two ways of getting a local copy of a configuration hosted on MOSRS:

- [Local-only copy](#rosie-checkout)
- [Local and remote copy (new remote configuration)](#rosie-copy)

#### Local-only copy {: #rosie-checkout }

```
rosie checkout <suite-id>/<branch>
```

where `<suite-id>` and `<branch>` are specific to the chosen model configuration and can be found in the respective [Run a Model](/models/run_a_model/) documentation. This creates a local copy of the configuration, which is placed in the `~/roses/<suite-id>` folder.

!!! tip
    To copy from the default branch (`trunk`), omit the `/<branch>` portion of the command.

> Configurations obtained in this way cannot be pushed back to the remote, so use of this command is recommended for testing and examining configurations.

Reviewer 2: ```suggestion
Configurations obtained in this way cannot be pushed back to the remote. Therefore, the use of this command is recommended for testing and examining configurations.
```

```
rose suite-gcontrol &
```

!!! tip
    The `&` is optional. It detaches the invoked process, allowing the terminal prompt to remain active while the GUI is open.

By default, the configuration, log files and outputs are copied to `/scratch/<project>/${USER}/cylc-run/<suite-id>`. See the respective [Run a model](https://docs.access-hive.org.au/models/run_a_model/) documentation for details on what outputs are generated and where to find them.

## Edit the model configuration

To edit the model configuration, run the following command from the configuration directory:

```
rose edit &
```

This opens the _Rose_ GUI that contains information about the configuration settings. For a description of the simple configuration settings, see the respective [Run a model](https://docs.access-hive.org.au/models/run_a_model/) documentation.

Once settings have been modified in the _Rose_ GUI, save them by clicking on the _Save_ button ![Save button](/assets/run_access_cm/save_button.png){: style="height:1em"}.

You are now ready to a run a model with _Rose/Cylc_.

[Run ACCESS-rAM](/models/run_a_model/run_access-esm){: class="text-card"}

> [Run ACCESS-CM](/models/run_a_model/run_access-cm){: class="text-card"}

Reviewer 2: ```suggestion
[Run an ACCESS model](/models/run_a_model/){: class="text-card"}
```
